src/main/java/com/amazon/ion/impl/_Private_IonReaderBuilder.java

```diff
@@ -275,11 +281,15 @@ static IonReader buildReader(
         if (source == null) {
             throw new NullPointerException("Cannot build a reader from a null InputStream.");
         }
+        int maxHeaderLength = Math.max(
+            _Private_IonConstants.BINARY_VERSION_MARKER_SIZE,
+            builder.getInputStreamInterceptors().stream().mapToInt(InputStreamInterceptor::headerMatchLength).max().orElse(0)
+        );
         // Note: this can create a lot of layers of InputStream wrappers. For example, if this method is called
         // from build(byte[]) and the bytes contain GZIP, the chain will be SequenceInputStream(ByteArrayInputStream,
         // GZIPInputStream -> PushbackInputStream -> ByteArrayInputStream). If this creates a drag on efficiency,
         // alternatives should be evaluated.
-        byte[] possibleIVM = new byte[_Private_IonConstants.BINARY_VERSION_MARKER_SIZE];
+        byte[] possibleIVM = new byte[maxHeaderLength];
         InputStream ionData = source;
         int bytesRead;
         try {
```

Comment:
Dosen't this need to read in a loop, to ensure sufficient bytes are present?

`read` can return any number of bytes, for any reason, and almost always needs to loop. I think this should loop until it either reads `maxHeaderLength` or hits EOF.

Looking at the comment below, this feels already-defective in a way that'll probably fail very rarely, but given support for additional encoding formats, it's now more worrisome.

Reply:
That's a good call. It's an easy enough fix, so I included it in the next revision. This doesn't necessarily solve the problem of an InputStream that is actually growing, but it does solve the problem of an InputStream that provides fewer bytes than are actually available when `read` is called.

Reply:
I think it "solves" the growing-stream situation by blocking until we can apply all potentially-relevant interceptors. That seems like the right behavior.